core/src/Streamly/Internal/FileSystem/Path.hs

```diff
@@ -1,515 +1,194 @@
-{-# LANGUAGE TemplateHaskell #-}
-
 -- |
 -- Module      : Streamly.Internal.FileSystem.Path
 -- Copyright   : (c) 2023 Composewell Technologies
 -- License     : BSD3
 -- Maintainer  : streamly@composewell.com
 -- Portability : GHC
 --
--- Well typed and flexible file systems paths, preserving the OS and filesystem
--- encoding.
---
--- You can choose the level of type safety you want. 'Path' is the basic path
--- type which can represent a file, directory, absolute or relative path with
--- no restrictions. Depending on how much type safety you want you can choose
--- appropriate type wrappers to wrap 'Path'. @File Path@ mandates the path to
--- be a file whereas @Abs (File Path)@ mandates it to be an absolute path
--- representing a file.
---
--- You can upgrade or downgrade the safety. Whenever a less restrictive path
--- type is converted to a more restrctive path type the conversion involves
--- checks and it may fail. However, a more restrictive path type can be freely
--- converted to a less restrictive one.
---
--- See the @streamly-filepath@ package for interworking with the 'OsPath' type.
--- The 'Path' type can be  converted to and from 'OsPath' type at zero cost
--- since the underlying representation of both is the same.
-
--- Conventions: A trailing separator on a path indicates that it is a directory.
--- However, the absence of a trailing separator does not convey any
+-- = User Notes
+--
+-- Well typed, flexible, extensible and efficient file systems paths,
+-- preserving the OS and filesystem encoding.
+--
+-- /Flexible/: you can choose the level of type safety you want. 'Path' is the
+-- basic path type which can represent a file, directory, absolute or relative
+-- path with no restrictions. Depending on how much type safety you want, you
+-- can choose appropriate type wrappers or a combination of those to wrap the
+-- 'Path' type.
+--
+-- The basic type-safety is provided by the
+-- "Streamly.Internal.FileSystem.PosixPath.LocSeg" module. We make a distinction
+-- between two types of paths viz. locations and segments. Locations are
+-- represented by the @Loc Path@ type and path segments are represented by the
+-- @Seg Path@ type. Locations are paths pointintg to specific objects in the
+-- file system absolute or relative e.g. @\/usr\/bin@, @.\/local\/bin@, or @.@.
+-- Segments are a sequence of path components without any reference to a
+-- location e.g. @usr\/bin@, @local\/bin@, or @../bin@ are segments. This
```

Comment:
A location is like an initial point and a segment is a route to some other place.

The examples are confusing. You can probably remove the examples. From a relative path, you cannot tell if it's a location or a segment. Location or a segment is what you already define. So the examples have no point?

Reply:
You can always tell. A location points to specific place, a location either starts with `/` or with `./` everything else is a segment. A location is either relative to root or explicitly relative to the current directory. `local/bin` is not relative to anything, it is a segment. `../bin` is not relative to anything, it is a segment.